app/lib/identity_tijuana/member_sync.rb

```diff
@@ -399,59 +399,66 @@ def self.sync(user, member, sync_type)
-              IdentityTijuana::Unsubscribe.create!(
+            user = User.new
+            user.created_at = member&.created_at # Preserve ID creation date
+            user_created = true
+          end
+
+          fields_updated = tj_changes.keys.dup
+          tj_changes.each do |key, value|
+            case key
+            when :email
+              if value.blank?
+                Rails.logger.warn("Member #{member&.id}, cannot erase the email address of an existing TJ user")
+                next
+              else
+                user.write_attribute(key, value)
+              end
+            when :postcode
+              user.postcode = Postcode.find_by(number: value)
+            when :state, :reason
+              # For state, since it is derived from the postcode in TJ,
+              # it can't mirror any changes made in ID
+              fields_updated.delete(key)
+            when :country
+              user.write_attribute(:country_iso, value.nil? ? nil : value[0..1])
+            when :is_member
+              user.write_attribute(:is_member, value)
+              msg = "#{value ? 'Subscribed' : 'Unsubscribed'} via update from Identity."
+              uae = IdentityTijuana::UserActivityEvent.create!(
+                :activity => value ? :subscribed : :unsubscribed,
                 :user => user,
-                :specifics => tj_changes[:reason] || msg,
-                :created_at => uae.created_at
+                :public_stream_html => msg,
               )
+              unless value
+                IdentityTijuana::Unsubscribe.create!(
+                  :user => user,
+                  :specifics => tj_changes[:reason] || msg,
+                  :created_at => uae.created_at
+                )
+              end
+            else
+              user.write_attribute(key, value)
             end
-          else
-            user.write_attribute(key, value)
           end
-        end
-        if user_created
-          user.save!
-          MemberExternalId.create!(member: member, system: 'tijuana', external_id: user.id) if user_created
-          Rails.logger.info("TJ user #{user.id} created from ID member #{member&.id}")
-        elsif fields_updated.count > 0
-          user.save!
-          Rails.logger.info("TJ user #{user.id} updated from ID member #{member&.id}: #{fields_updated.join(' ')}")
+
+          if user_created
+            user.save!
+            MemberExternalId.create!(member: member, system: 'tijuana', external_id: user.id)
+            Rails.logger.info("TJ user #{user.id} created from ID member #{member&.id}")
+          elsif fields_updated.count > 0
+            user.save!
+            Rails.logger.info("TJ user #{user.id} updated from ID member #{member&.id}: #{fields_updated.join(' ')}")
+          end
+        rescue StandardError => e
+          Rails.logger.error "Failed to sync (#{sync_type}) member id:#{member&.id} to TJ, error: #{e.message}"
+          raise e
```

Comment:
Should this be happening in the the two call sites above (i.e. `export_member` & `import_user`) - that way we can print an error with the sync id, type and direction, and we catch /any/ exception that gets raised in `sync`?

Reply:
This is strange – your comment appears as `5 hours ago`, but that code already changed yesterday at 3pm (??). This commit is already [propagating sync id and type](https://github.com/GetUp/identity-tijuana/pull/69/commits/865c04741dd505f91e4762d4945f4af37cd3a948).

Reply:
Yeah dunno, GH's going downhill fast post-MS-buyout.

I was thinking about adding the `rescue`, not about the logging though - since any code in the `sync` method could raise an exception, shouldn't the callers be doing the rescuing and logging?

Reply:
Yes, I agree, but there are a few exceptions. `Sync type` is not known to the callers, unless we want to throw exception from the sync methods ie. `export_member` and `import_user`. But then we might as well handle it in-situ.

A minor thing was that all logging was happening in the `sync` service, so just thought to keep the same approach. Still think this may be a good approach.